[v11] Add in app access connection monitoring.

mdwn opened the pull request:
Backport #17436 to branch/v11

Commit:
Add in app access connection monitoring.

Application access connection monitoring has been introduced so that, when a
lock is created, application access connections will be interrupted until the
lock has been cleared. This includes web sockets and TCP applications.

Commit:
Close the async emitter on exit.

Commit:
Preserve http tracing during initial auth failure.

Commit:
Roman's feedback.

Commit:
Linting issues.

Commit:
Roman's feedback.

r0mant reviewed (dismissed):
We should not merge these backports until the flakiness of new tests is fixed as shown in https://github.com/gravitational/teleport/issues/17864#issuecomment-1306009298.

mdwn committed:
Make TestTCP* tests in appaccess more deterministic. (#18233)

The appaccess TestTCP* tests are highly reliant on time. This has been
reduced (but not eliminated) by using a fakeClock and a channel for
signaling monitor triggered connection closures.
